Add a positions object with a parsed selection criterion

GiudGiud opened the pull request:
Thoughts on this?

Seems like it could be useful. But it does not quite do what the user wanted (material properties) here: https://github.com/idaholab/moose/discussions/30765
It's inspired from that need though.

Edit: actually using the auxvariable output of the property, this does work for matprops 

name could be ElementParsedSelectionPositions instead since it keeps the element centroid positions
or we could add (in the future) an argument to decide the type of loops (nodal, Qps, faces, ...)

GiudGiud committed:
Ban material properties from nodal parsed-aux

GiudGiud committed:
Add a positions object for selecting elements based on a parsed expression
refs #23587

GiudGiud committed:
Add test and documentation
refs #23587

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2952242/), step Docs: sync website on 775b5a7 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30767/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2952291/), step Generate coverage on 775b5a7 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>f0af66</th>
      <th colspan=4>#30767 775b5a</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.47%</td>
      <td>85.48%</td>
      <td>+0.00%</td>
      <td>89.61%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>114769</td>
      <td>114838</td>
      <td>+69</td>
      <td>69</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19506</td>
      <td>19514</td>
      <td>+8</td>
      <td>8</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30767/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30767/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30767/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 89.61% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad commented on framework/src/positions/ParsedSelectionPositions.C:
```diff
@@ -0,0 +1,168 @@
+    if (std::find(_xyzt.begin(), _xyzt.end(), name) != _xyzt.end())
+      paramError(
+          "functor_names",
+          "x, y, z, and t cannot be used in 'functor_names'. Use functor_symbols to disambiguate");
```
The second sentence of this error message seems misleading. Seems like you should just delete the second sentence?

GiudGiud replied:
second sentence is telling the user how to actually do this if needed

i ll try to reword

lindsayad commented on framework/src/positions/ParsedSelectionPositions.C:
```diff
@@ -0,0 +1,168 @@
+      variables += (variables.empty() ? "" : ",") + name;
+
+  // xyz and t are likely useful here
+  for (auto & v : _xyzt)
+    variables += (variables.empty() ? "" : ",") + v;
+
+  // base function object
+  _func_F = std::make_shared<SymFunction>();
+
+  // set FParser internal feature flags
+  setParserFeatureFlags(_func_F);
+
+  // parse function
+  if (_func_F->Parse(_expression, variables) >= 0)
+    mooseError("Invalid function\n", _expression, "\nError:\n", _func_F->ErrorMsg());
+
+  // optimize
+  if (!_disable_fpoptimizer)
+    _func_F->Optimize();
+
+  // just-in-time compile
+  if (_enable_jit)
+  {
+    // let rank 0 do the JIT compilation first
+    if (_communicator.rank() != 0)
+      _communicator.barrier();
+
+    _func_F->JITCompile();
+
+    // wait for ranks > 0 to catch up
+    if (_communicator.rank() == 0)
+      _communicator.barrier();
+  }
+
+  // reserve storage for parameter passing buffer
+  _func_params.resize(_n_functors + 4);
+
+  // keep pointers to the functors
+  for (const auto & name : _functor_names)
+    _functors.push_back(&getFunctor<Real>(name));
```
This kind of logic is repeated in a fair number of places
```
(moose-python) lindad@rod:/data/lindad/projects/moose/framework(avoid-refine-in-reinit-noop-30725)$ git grep --name-only JITCompile
include/utils/ADFParser.h
src/auxkernels/ArrayParsedAux.C
src/auxkernels/ParsedAux.C
src/auxkernels/ParsedVectorAux.C
src/materials/DerivativeParsedMaterialHelper.C
src/postprocessors/ParsedPostprocessor.C
src/scalarkernels/ParsedODEKernel.C
src/utils/ADFParser.C
src/utils/FunctionParserUtils.C
```
Can we make a helper function somewhere to help reduce the number of times we repeat this pattern? There is certainly some variation throughout these grep match occurrences but maybe the helper function could take arguments that support those variations

lindsayad commented on framework/src/positions/ParsedSelectionPositions.C:
```diff
@@ -0,0 +1,168 @@
+ParsedSelectionPositions::initialize()
+{
+  clearPositions();
+  const auto n_points = _fe_problem.mesh().nElem();
```
Would @DanielChou0916 be happy with just gathering the positions of the elements (represented by their centroids) as opposed to the actual integration points?

lindsayad commented on framework/src/positions/ParsedSelectionPositions.C:
```diff
@@ -0,0 +1,168 @@
+      _func_params[i] = (*_functors[i])(elem_arg, state);
+
+    // Positions and time
+    const auto & centroid = elem->true_centroid();
```
```suggestion
    const auto centroid = elem->true_centroid();
```

lindsayad commented on framework/src/positions/ParsedSelectionPositions.C:
```diff
@@ -0,0 +1,168 @@
+
+    // Positions and time
+    const auto & centroid = elem->true_centroid();
+    for (const auto j : make_range(LIBMESH_DIM))
```
```suggestion
    for (const auto j : make_range(Moose::dim))
```

GiudGiud commented:
git s buggy so replying here

: 
We already have a quadraturePointPositions btw. We could add an option to this ParsedPositions to be able to down-select from an existing Positions object .

> Otherwise it feels a bit like adding code just because we can

Every system ought to have parsed capabilities imo. It's just so much more flexible and helps no-code users of our software.

GiudGiud commented:
> Can we make a helper function somewhere to help reduce the number of times we repeat this pattern? There is certainly some variation throughout these grep match occurrences but maybe the helper function could take arguments that support those variations

seems tough with the number of attributes that may not be shared by all these objects. I ll take a look. I think all these attributes can then just be parameters of the utility, along with a bunch of booleans to enable/disable things

lindsayad commented:
Sounds good

GiudGiud committed:
CHange the parsed Position concept to a parsed down selection
Letting us handling any type of input positions

lindsayad reviewed (commented):
I really like it. Great use of `ElemPointArg`. I added some sanity checking in the commit I just pushed. Let me know how you feel about them.

lindsayad commented on framework/src/positions/ParsedDownSelectionPositions.C:
```diff
@@ -167,6 +167,7 @@ ParsedDownSelectionPositions::initialize()
   for (const auto pos_ptr : _positions_ptrs)
     n_points += pos_ptr->getNumPositions(initial);
   _positions.reserve(n_points);
+  mooseAssert(_communicator.verify(n_points), "We should have a synchronized number of points");
```
If you think this is a no-brainer, then feel free to remove it

lindsayad commented:
Well the testing is failing in parallel because of ghosting issues. Instead of adding my semiverify check that can be avoided by just evaluating for a given element ` if (elem->processor_id() == this->processor_id())`. Then there are no ghosting issues, you only evaluate the function on one rank for a given element, and then logically there is no chance for a disagreement between ranks

GiudGiud committed:
Address Logan's review of #30737: supress param on tag aux

GiudGiud committed:
Adapt tests post suppression of tag aux execute_on parameter

GiudGiud committed:
Make downselection more inclusive: if any process says yes, keep it
Support using nodes in downselection which might have multiple elements meeting or not meeting the restriction

GiudGiud committed:
Move common code between parsed expression objects into the util

GiudGiud reviewed (commented):
looks like it s letting me submit now

GiudGiud commented:
Added the sweeps so we dont get surprised

GiudGiud commented on framework/src/positions/ParsedDownSelectionPositions.C:
```diff
@@ -0,0 +1,226 @@
+
+        // Keep points matching the criterion
+        if (value > 0)
+          keep_positions[i_pos - 1] = 2;
```
Yeah I did to make keeping the default. It works better for a potential node at a process boundary

GiudGiud replied:
fixed the comment

GiudGiud committed:
Add a counter for C++ vibes
Use an enum to show the selection decisions
Improve comments on what the inclusion of positions means

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

lindsayad commented on framework/src/positions/ParsedDownSelectionPositions.C:
```diff
@@ -0,0 +1,227 @@
+        if (value > 0)
+          keep_positions[i_pos] = PositionSelection::Keep;
+        // Dont exclude a point we already chose to keep. This 'inclusivity'
+        // means that a position at a node on the edge of a block gets included
```
```suggestion
        // means that a position at an interface between elements can get included if the functor evaluates greater than 0 for any of the elements
```

GiudGiud commented:
Thanks for the review!
